File: Python/530. Minimum Absolute Difference in BST.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    
    def __init__(self):
        self.ans = float('inf')
        self.val = -1
        
    def getMinimumDifference(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        def dfs(node):
            if node.left:
                dfs(node.left)
            if self.val >= 0:
                self.ans = min(self.ans, node.val - self.val)
            self.val = node.val
            if node.right:
                dfs(node.right)
            return self.ans
        return dfs(root)

File: C++/530. Minimum Absolute Difference in BST.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int ans = INT_MAX, val = -1;
    int getMinimumDifference(TreeNode* root) {
        if(root->left)
            getMinimumDifference(root->left);
        if(val >= 0)
            ans = min(ans, root->val - val);
        val = root->val;
        if(root->right)
            getMinimumDifference(root->right);
        return ans;
    }
};